### ISSUE TITLE
[Feature] Samsung SmartThings API support for Input Source, etc.

### ISSUE THREAD

**Issue by HVR88** [tavicu/homebridge-samsung-tizen#321]
Add  (additional) support for the SmartThings API, which will allow direct/instant query and setting of Input Source among other features not possible with the remote control API being used right now. It will solve a number of the shortcomings listed on the first page of the Wiki, including: checking current channel, volume, mute status, source selection.

I briefly mentioned this in another post, but I'm creating a dedicated feature request for it now after finding multiple code examples where this feature was added to previous integration that, like homebridge-samsung-tizen, were using the V2 API or legacy remote control APIs exclusively. In short: the developers were able to integrate SmartThings support without refactoring their whole projects. ;)

**Similar integrations/drivers for other platforms, using v2 remote API, SmartThings API and legacy APIs:**

**Hubitat**: https://github.com/DaveGut/HubitatActive/blob/master/SamsungTvRemote/Samsung%20TV%20Remote.pdf

**Homey**: https://github.com/balmli/com.samsung.smart

**Home Assistant**: https://github.com/jaruba/ha-samsungtv-tizen

**Comment by tavicu** [comment]
Hi @HVR88, Thank you for providing info. I really appreciate when someone want's to help and try to do things better.
I know the SmartThings api will offer us more functionality and less room for errors.

The biggest problem migrating to SmartThings is time. I have a full work job and don't have a lot of free time. And the time i have i want to spent it with my family :)

After i finish the Art Mode support for Frame TVs i will look over the SmartThings API. The problem is that i don't know to say if it will be implemented or how much time it will take.

**Comment by HVR88** [comment]
Thanks!

I hear you.  If this was last year, when I had a lot of time, I'd look at this and try to implement it and then issue a pull request. :) Right now I'm also generally working long hours and it's all I can do to spend a little bit of time just setting up Homekit and my devices. :)

**Comment by tavicu** [comment]
I have looked over the SmartThings API and play with it. I don't like it (sometimes it tells me that TV is off and can't send commands when it's on ...) but it will not be hard to implement the functionality to switch to an input.

After i finish with Frame TV implementation i will take a look.

**Comment by mrMiimo** [comment]
Last night I also played with the https://github.com/jaruba/ha-samsungtv-tizen plugin (based on smartthing api, I believe) integrating HomeAssistant with HomeKit.

I solved the problem of updating the status ON / OFF I changed the update_method (from "ping" to "websocket").
I've seen that switching TV channels and inputs is actually more useful with API.

I guess it will take time, unfortunately I'm not a developer ...

**Comment by tavicu** [comment]
> Last night I also played with the https://github.com/jaruba/ha-samsungtv-tizen plugin (based on smartthing api, I believe) integrating HomeAssistant with HomeKit.

I have looked and i don't find their method very good as performance and reliable.
They are sending a command to TV and wait for the answer. if the TV will not answer then means the websocket is down and that also means the TV is off.

But sometimes websocket takes longer to respond, if you add a timeout it may give you false results. Ping on the other hand, it responds right away and have nothing to do with the OS of the TV.
If websocket was down right away when the TV entered sleep mode, yeah, i would thing it's a viable option. But websocket stay active even in sleep mode, so not much help over here :)

I have 4 Samsung TV and none of them returned an invalid response to ping.

Also ha-samsungtv-tizen "ping" method is not ping ... it's actually sending a http request to `/api/v2/`. We do the same, but only if TV responds to ping and we also cache the request for 2 seconds so we don't stress the TV with requests.

Unfortunately what is working for you, it's not for others. And with my plugin i want to cover as much TV's i can and without advanced configs. Y

ou can see why i want to do this by looking at this issue https://github.com/tavicu/homebridge-samsung-tizen/issues/323 where the message tells him to check the Wiki for more details but he opens an issue. I respond him with the link to Wiki where everything is detailed where is say he's tv is not compatible and that Samsung removed the API and he ask if it's going to be fixed :|

**Comment by mrMiimo** [comment]
Clear as water.

> You can see why i want to do this by looking at this issue #323 where the message tells him to check the Wiki for more details but he opens an issue. I respond him with the link to Wiki where everything is explained in details (is say he's tv is not compatible and that Samsung removed the API) and he ask if it's going to be fixed :|

I read #323 this morning 🤦🏻‍♂️  I thought: "it's hard to maintain a partnership like this community if people don't want to read and just pretend"

**Comment by tavicu** [comment]
> I read #323 this morning 🤦🏻‍♂️ I thought: "it's hard to maintain a partnership like this community if people don't want to read and just pretend"

Yup ... that's why I didn't do much in the last time. Cause all i've seen where issues like that! But some of you started to help me with info and that shows me you guys care. That's why i changed the way apps are opened, started implementing Frame Support and next look over the SmartThing API. Because you guys deserve it by being polite and do more than "i want that", "it's not working" :)

i will always help someone who is polite and do the minimum research before asking for something!

**Comment by tavicu** [comment]
A small update:
Today i played with smathings api! It's looking good :)

After i finish with Frame i will implement selecting an input with smartthings. I like that you can select Live TV :)

**Comment by HVR88** [comment]
That's fantastic news. As I mentioned in another thread, today, your plugin is the most stable right now. I've tried a lot of projects that are supposed to communicate with Samsung TVs in the past month (not for Homekit) and all the others are broken in some way or another right now.

**Comment by tavicu** [comment]
Yup ... that's why i started this plugin in the first time :D none of them was good :P 

And that's why i don't want to implement things that "may" work ... i only do things that DO work for sure and stable :) 
The plugin is made for everybody, not just for those who knows how to code something :D 

Eh, i will try to get more informations regarding the smarthings api and if any other thing beside selecting an input looks stable and nice to implement it, i will do it :D

**Comment by deepontech** [comment]
Thanks for improving this amazing tool.

**Comment by town3r** [comment]
Any update on being able to select a Source Input?

Finally figured out how to add the apps as inputs but would love a Live TV & Apple TV box Source Input option.

**Comment by tavicu** [comment]
Hei @town3r , no news yet. I was very busy at my work and didn't had time to breath :D 

But I expect things to start to calm down being the holiday period.

So the feature is on my priority list as soon as i will have some days off.

**Comment by pseudohero** [comment]
Very excited for this functionality to be added!

**Comment by jakubheglas** [comment]
+1

**Comment by tavicu** [comment]
Hi. Small update here. I'm very excited for this too :))

I will have some time this month and i will do my best to implement it asap.

**Comment by tavicu** [comment]
Ok, released the first beta :)

I've added the option to set an input/switch to open a source using SmartThings API.
Right now the input will not remain active. I still have to make some checks and see if this is possible because the SmartThings api when you are on an app it will return as active source the last input you were on.

For example if you were on `HDMI1` and open `YouTube` the endpoint will respond with the active source still as `HDMI1` :)

Also, make sure you set your IDs ok because the plugin does not return any error now if it's not working. It will just ignore the command.

***

You will need the API Key for the SmartThings API and Device ID.

How to get the informations.

**First make soure your TVs are added in SmartThings App**

**API Key**
1. Go to https://account.smartthings.com/tokens
2. Login with the same account you are logged in SmartThings App on your device
3. Generate a new token

**Device ID**
1. Go to  https://account.smartthings.com/
2. On top bar select `My Devices`
3. From the list select your TV
4. The Device ID will be part of the page URL
5. Example url: `https://[...].api.smartthings.com/device/show/XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXX`
6. Device ID will be: `XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXX`

***

Time to install the beta version:

### If you are using Config Ui X (i think it should run on docker but i don't know for sure) it's the easiest way!
1. Log in to Config Ui X
2. Go to Plugins
3. Click on the wrench and select `Install Previous Version`
![Screenshot 2021-03-14 at 17 03 32](https://user-images.githubusercontent.com/172068/111073741-81ce6b80-84e8-11eb-94d0-085a654fb5cb.png)
4. In the dropdown make sure it's selected the beta version: `v5.2.0-beta.0 - beta` and click `Install`.
![Screenshot 2021-03-14 at 17 05 03](https://user-images.githubusercontent.com/172068/111073760-a32f5780-84e8-11eb-9fe8-ba9ec4a101cd.png)

### For those who don't have Config Ui X or don't see the `Install Previous Version` option 
You can install the plugin with `npm install` command like this: `npm install homebridge-samsung-tizen@beta`. If you have the plugin installed globally don't forget the `-g` attribute. Like this: `npm install homebridge-samsung-tizen@beta -g`

***

Time to update the config.

Possible values for now: `digitalTv`, `USB`, `HDMI1`, `HDMI2`, `HDMI3`, `HDMI4`, `HDMI5`, `HDMI6`

Example of config:
```
{
    "platform": "SamsungTizen",
    "devices": [
        {
            "name": "Bedroom TV",
            "ip": "10.20.30.40",
            "mac": "A0:B1:C2:D3:E4:F5",
            "api_key": "0d055d9b-d662-42a5-ae25-dfe254f10c09",
            "device_id": "5d9967aa-c411-4e12-a998-c4ec48754f08",
            "inputs": [
                {
                    "type": "input",
                    "name": "Live TV",
                    "value": "digitalTv"
                },
                {
                    "type": "input",
                    "name": "Apple TV",
                    "value": "HDMI1"
                }
            ],
            "switches": [
                {
                    "name": "Switch Live TV",
                    "input": "digitalTv"
                }
            ]
        }
    ]
}
```

**Comment by ManuISEN** [comment]
Hi @tavicu,

Tested today but have issues.

I'm running homebridge on a raspberrypi 3, installed v5.2.0-beta.0 from Config Ui X.

My config.json :

	{
		"platform": "SamsungTizen",
		"app_list": true,
		"devices": [
			{
				"name": "TV",
				"ip": "10.20.30.40",
				"mac": "A0:B1:C2:D3:E4:F5",
				"api_key": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXX",
				"device_id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXX",
				"keys": {
					"ARROW_UP": "KEY_UP",
					"ARROW_DOWN": "KEY_DOWN",
					"ARROW_LEFT": "KEY_LEFT",
					"ARROW_RIGHT": "KEY_RIGHT",
					"SELECT": "KEY_ENTER",
					"BACK": "KEY_RETURN",
					"PLAY_PAUSE": "KEY_PLAY_BACK",
					"INFORMATION": "KEY_INFO"
				},
				"inputs": [
					{
						"name": "Apple TV 4K",
						"type": "input",
						"value": "HDMI3"
					},
					{
						"name": "Xbox One X",
						"type": "input",
						"value": "HDMI4"
					},
					{
						"name": "TNT",
						"type": "input",
						"value": "digitalTv"
					}
				]
			}
		]
	}

In Home app, I can't change inputs, it does nothing to the TV.
In homebridge logs, I have this error message:

[07/10/2021, 15:45:29] [TV] Cannot read property 'send' of undefined

Note: I'm also running PiHole to block adds, but don't see something blocking related to homebridge/SmartThings API calls in its logs.

**Comment by ManuISEN** [comment]
Set `app_list` to `false` fixed the websocket exception :)

Changing inputs is working 👍

**Comment by rfiorentino1** [comment]
For some reason, none of the keys are working for me. Maybe I need to enable something for this to work? Everything else, on/off, inputs, etc works great! But left/right or numbers like KEY_1 don't do anything. Anyone else having this issue?
Thank you so much for all of your work, it's much appreciated, and now allows me to control the TV through Alexa as well, via [HomeBridge Alexa](https://github.com/NorthernMan54/homebridge-alexa#readme)

**Comment by tavicu** [comment]
> For some reason, none of the keys are working for me. Maybe I need to enable something for this to work? Everything else, on/off, inputs, etc works great! But left/right or numbers like KEY_1 don't do anything. Anyone else having this issue?

please paste here the inputs part of the config :)

**Comment by rfiorentino1** [comment]
I've added it as a switch, instead of an input. Is this why? I've also done it through the Homebridge UI, instead of manual config, but here's the switches part of the config:
"switches": [
                        {
                            "name": "Comcast",
                            "power": true,
                            "input": "HDMI2"
                        },
                        {
                            "name": "Apple",
                            "power": true,
                            "input": "HDMI1"
                        },
                        {
                            "name": "Fox News",
                            "power": true,
                            "input": "HDMI2",
                            "command": [
                                "KEY_1,KEY_1,KEY_1,KEY_0,KEY_ENTER"
                            ]
                        }
                    

And here's the inputs section:

"inputs": [
                        {
                            "type": "input",
                            "name": "Comcast",
                            "value": "HDMI2"
                        },
                        {
                            "type": "input",
                            "name": "Apple TV",
                            "value": "HDMI1"
                        }
                    ],

**Comment by ramm64** [comment]
Hi there — Just a quick note to report that I jumped into the beta release as I wanted to test out using the Samsung SmartThings API to control my TV. I went ahead and installed it today and played around with the settings. A couple of observations to offer so far:

- I have an Apple TV connected to HDMI 1. Since I have CEC switched on, the Apple TV turns on when the TV is turned on and vice versa. However, after adding HDMI 1 as an input in the plug-in, switching to that input does not power on the Apple TV. The TV eventually displays a screen saying “no power, push this button to power your device” and then the Apple TV wakes up.
- I also have a cable TV box, connected to HDMI 2. I normally don’t watch cable, so I only have a few channels I’m interested in. So I decided to use the plug-in to create a few switches with commands that would tune in these channels. At first I used keys for this, but they didn’t work. I next entered values in the Channel Number field in settings and … that worked! I didn’t expect the channel number values to work on externally-attached cable boxes, so I am pleased.

Overall I am very happy with how this works so far. Looking forward to the next release! Thank you for outputting such a stable and useful plug-in. 🙏🏽

**Comment by tavicu** [comment]
Hi @ramm64, thank you for the feedback.

1. I will take a look at this but I don't think I have what to do because I just call an samsung api that switch the input and after that everything it's on Samsung's part.
2. Yup, the channel actually sends a list of commands. For example channel: 42 will send `KEY_4, KEY_2, KEY_ENTER` so it should work anywhere regarding on what input/app you are :)

***

As an update of the feature, November it's a very busy period for me so i didn't have time to do anything. I will resume working this month and i hope to release it at the middle of January :)

I also want to update the Wiki before this release so that will take some time

**Comment by HVR88** [comment]
I finally had a chance to test this myself - life gets in the way.

Input switching seems to work very well and quickly, however there's a problem that I can't get around...

Using the device in a Homekit scene. If the scene must turn the TV ON and switch inputs, the input switch doesn't happen.

For example:

Plugin is configured with various, including HDMI1 and HDMI2
I manually set my TV to HDMI1 and then turn it off - works great from the Home app
Create a scene an include ONLY the TV - set the scene to put the TV ON and Input to HDMI2
Close/save the scene
Trigger the scene - the TV will turn on with HDMI and never switch inputs to HDMI2

**Comment by tavicu** [comment]
Update. I have released the second beta (**v5.2.0-beta.1**)
For installing and updating follow the steps from here: https://github.com/tavicu/homebridge-samsung-tizen/issues/321#issuecomment-935886172

What changed:
- Interpretate error messages from SmartThings API
- Add option in config ui to set the key and id.
- Throw errors if **api_key** and **device_id** are not set. Still work in progress cause i want to add link to Wiki page
- Check and selects active inputs
- And other things that are not related to this implementation.

I will say the feature is 95% done.

I still have some things that i want to change before release (removing the feature to list installed applications for example) and i will also want to improve the Wiki section.

So unfortunately, even if this feature will be 100% done it will not be released until the documentation will be ready. (At least this is my position right now, it may change).

**Comment by HVR88** [comment]
I believe the solution is that an input delay needs to be baked-in automatically when an input change follows a power state change. This is a common parameter in IR control.

This same issue has affected other Homekit plugins, such as Denon-TV that controls Denon and Marantz AVRs. In fact I just reported a regression of this here which has been fixed: https://github.com/grzegorz914/homebridge-denon-tv/issues/129#issuecomment-1034573872

**Comment by tavicu** [comment]
> I believe the solution is that an input delay needs to be baked-in automatically when an input change follows a power state change. This is a common parameter in IR control.

Setting an automation that turn the tv ON then change the input will not result in a single command to the plugin that i will know to interpret. It will be two commands.

Right now the only way i'm thinking is adding a delay to Power On but i don't want to do this since it will affect everybody. So implementing a problem for 100% of the users to resolve one small issue it's not ok :)

I will think and test a few things. First i must see after which time the TV will accept the Input switch after the TV turned on.

**Comment by HVR88** [comment]
Take a look at the source for the plug-in in linked. There is no delay as far as I know on regular input changes. Only on power on sequencing. It might be necessary to add the delay to the power on action - that would benefit everyone because no command can immediately follow a power command workout the tv being ready to accept it.

**Comment by tavicu** [comment]
> Take a look at the source for the plug-in in linked. There is no delay as far as I know on regular input changes. Only on power on sequencing. It might be necessary to add the delay to the power on action - that would benefit everyone because no command can immediately follow a power command workout the tv being ready to accept it.

Adding delay to power on command will not solve the problem. I just tested it and Apple don't wait for Main Accessory to finish responding before sending the change input. It just sends both commands at the same time.

So I have to implement something in the change input action to wait before calling the API if the TV was just turned ON!

**Comment by HVR88** [comment]
I think the plug-in might be best served with the delay as a variable in the ui. This is what Logitech and other remote manufacturers have in place for their products to be able to deal with customers different systems. 

If apples HomeKit was natively scriptable I’d make my own delay, but that’s unfortunately not possible. I’d have to move the tv control over to a different platform or the entire automation sequence.

**Comment by HVR88** [comment]
All decent home automation platforms are fully scriptable. Indigo, Home Assistant, etc. 

I’ve never seen any delay options in home.app. And nothing of any kind that is able to solve this basic requirement of turning on the tv to the correct input.

**Comment by jakubheglas** [comment]
@tavicu That is very elegant actually, thank you!

@HVR88 Just because you didn't see it doesn't mean it doesn't exist :)  You can, _for example_, make a catch-all automation that executes on TV On, waits 2 seconds and then triggers the input change. Quite easily right in Home.app, that is.

**Comment by tavicu** [comment]
In my opinion HomeKit is just plain dumb! It was in the first smart automation systems but they didn’t improve it by much! Just little by little every year!

SmartThings api for example, when you send a command you can send a list of commands, not just one! Let’s say turn on then change input!

All you have to do is to send the commands and the api will turn on the tv then wait until it can change the input.

My goal with this integration is to learn as much as i can regarding the SmartThings api and then start to switch functions to it step by step!

Unfortunately there are still some missing things for the older tvs. For example, you can’t turn on older tvs with smartthings api.

**Comment by HVR88** [comment]
@jakubheglas You have to convert the automation to a shortcut - lots of problem reports about dealing with shortcuts on Homepod - and Homepod is almost exclusively what I use with Homekit. My reason to use Homekit at all is for whole-home Siri support as I already have a very capable automation system that handles most of my devices like Z-Wave switches.

This plugin was also my main reason for setting up Homebridge - nothing else that I tested on other automation systems worked well with my 2020 TV. Thanks @tavicu for adding this feature and so quickly.

**Comment by HVR88** [comment]
@tavicu With 5.2.0beta.4 is it necessary to set the "delay" property or do you include a default delay for all inputs? And can you confirm if the delay property only applies to input commands after power commands?

**Comment by tavicu** [comment]
> @tavicu With 5.2.0beta.4 is it necessary to set the "delay" property or do you include a default delay for all inputs? And can you confirm if the delay property only applies to input commands after power commands?

A default delay of 2.5 seconds it’s applied by default only after a power switch command was sent.

That option is just in case it doesn’t work at it should!

**Comment by ramm64** [comment]
Hi @tavicu — Posting this here in case you have any insight. I’ve been using the beta plugin and it works well. However, one issue comes up related to the TV power state: I have an automation in the Home app to turn OFF some Hue lights if/when the Samsung TV turns OFF. Most of the time this works as expected, but many times the lights will turn OFF even though the TV continues to be ON. I can trace this issue to the automation itself, so somehow HomeKit thinks the TV has turned off (even though it’s still on) so my lights get turned off as well.

I set Homebridge to DEBUG mode but I see very few messages related to the plugin in the logs. I see the messages indicating that the lights have been switched off, but no messages from your plugin that would indicate the TV has powered off as a trigger to the automation being fired.

One message that I see in the log is this one:

`[16/02/2022, 12:34:51] [My TV] [DEBUG] https://api.smartthings.com/v1/devices/MY DEVICE ID/status undefined`

Does this ring a bell with you? Could this be the possible source of the issue? Any insight would be welcome. Thanks again for all your efforts with this plugin.

**Comment by tavicu** [comment]
Hi @ramm64 ,

i can help you adding a few lines of debug but please open a new issue since it’s not related to this feature!

thank you!

**Comment by ramm64** [comment]
> Hi @ramm64 ,
> 
> i can help you adding a few lines of debug but please open a new issue since it’s not related to this feature!
> 
> thank you!

Done! #474 created. Thanks again.

**Comment by NickDarvey** [comment]
@tavicu 
>A default delay of 2.5 seconds it’s applied by default only after a power switch command was sent.
>
>That option is just in case it doesn’t work at it should!

Does this option work with apps too? i.e.
```json
{ "inputs": [
    "type": "app",
    "name": "Spotify",
    "value": "3201606009684",
    "delay": 9000
] } 
```

**Comment by tavicu** [comment]
Yes, but this will be changed in the next beta. It will be something general for every command, not only inputs

**Comment by tavicu** [comment]
New beta released (**v5.2.0-beta.5**)
For installing and updating follow the steps from here: https://github.com/tavicu/homebridge-samsung-tizen/issues/321#issuecomment-935886172

What changed:

1. Add delay for all requests if the TV just turned ON! The default is set to 3s but this can be change with a new attribute `wait_time` set per device that will need to be set in milliseconds. This should not be changed and leave the default value but some older TVs a longer time is required.
``` 
"devices": [
    {
        ...
        "wait_time": 6000
    }
]
```

2. Add option to set a specific volume (0 - 100) with SmartThings API and custom Switches like this:
```
{
   ...
    "switches": [
        {"name": "Volume 5",  "volume": 5}
    ]
}
```

3. Add `race` for all requests. This means every command you send to TV will report back to HomeKit in maximum 2.5 seconds. Even if it takes longer to execute, it will execute in the background.
4. Cache the response from SmartThings API Status for 2.5 seconds (not administrable). This is implemented to not send too many request to SmartThings when we want to get the status for different variables (like active input).
5. If SmartThings API is set, use it to change channel. Until now a hack was used where we divided the channel number and send a list of commands. For example channel 55 we send `KEY_5, KEY_5, KEY_ENTER`
6. If Debug mode is turned ON, show every response from SmartThings API (this will probably be reverted before going live because it's a long response)


Under the hood there were many changes in the new beta even if some things works the same way. So please, if you find any error report back so we can investigate :)

Happy testing :)

**Comment by ramm64** [comment]
Installed Beta 5 and am now getting these messages for each of my custom switches:

`[Samsung TV BBC (164)@Samsung TV BBC (164)@On] The read handler for the characteristic 'On' didn't respond at all!. Please check that you properly call the callback!`
`[21/02/2022, 11:50:35] [homebridge-samsung-tizen] This plugin slows down Homebridge. The read handler for the characteristic 'On' didn't respond at all!. Please check that you properly call the callback! See https://git.io/JtMGR for more info.`

The switch simply powers on the TV, changes input to HDMI2 (cable TV), and changes the channel to the BBC.

**Comment by tavicu** [comment]
Possible the channel setting since i don’t know it’s working on HDMI! Could you try and remove it from the switch and let only power and input?

Also debug mode so we can see if there are errors

**Comment by tavicu** [comment]
> Installed Beta 5 and am now getting these messages for each of my custom switches:
> 
> `[Samsung TV BBC (164)@Samsung TV BBC (164)@On] The read handler for the characteristic 'On' didn't respond at all!. Please check that you properly call the callback!` `[21/02/2022, 11:50:35] [homebridge-samsung-tizen] This plugin slows down Homebridge. The read handler for the characteristic 'On' didn't respond at all!. Please check that you properly call the callback! See https://git.io/JtMGR for more info.`

@ramm64, Done, fixed it :)

**Comment by ramm64** [comment]
Wow, that was fast! 😃

Installed beta.6, restarted and can report no nasty errors in the logs. I'll do more testing this afternoon here (local time) and will leave you a report on how the new volume feature and other new additions are working for me.

Thank you -- and have a nice rest of your evening.

**Comment by ramm64** [comment]
Was able to do some testing today with beta.6. Here are my results:

- New "set volume" feature is working, but only if the TV output is set to the TV speaker. My audio is configured to output using the HDMI ARC port; **can this feature be supported to get volume set via HDMI?**
- New "set channels" feature is not working, at least for my cable TV box. Switches for my cable TV stations all have channel numbers and back in beta.4, the channels changed correctly. In beta.6, the presence of channel numbers in the "channel" parameter of the switch leads to instability, making the Home app report "not responding" status on my TV. Deleting the value from the "channel" parameter makes the plugin responsive again
- All switches created in this version "mis-behaved" in the Home app. For example, activating a switch to change channels to the BBC made switches for other channels come "on" at the same time (I believe this was due to every switch I configured having a value in the "input" parameter: HDMI2. Since the BBC is on HDMI2, all other switches with HDMI2 in the "input" parameter also got activated). In previous versions of the plugin, each switch in the Home app would act as a "toggle," that is, the switch would change to "on," the commands would be performed on the TV, and then the switch would go back to "off."

Here is a paste of the debug log showing the errors when I tried to activate a switch to change the channel:

```
[21/02/2022, 15:19:51] [Samsung TV] [DEBUG] https://api.smartthings.com/v1/devices/MY_TV_DEVICE_ID/commands {
  commands: [
    {
      component: 'main',
      capability: 'tvChannel',
      command: 'setTvChannel',
      arguments: [Array]
    }
  ]
}
2022-02-21T21:19:51.591Z HAP-NodeJS:EventedHTTPServer:Connection [::ffff:IP_ADDRESS] Sending HAP event notifications [ { aid: 1, iid: 352, value: 0 } ]
[21/02/2022, 15:19:51] [Samsung TV] [DEBUG] {"requestId":"MY_TV_DEVICE_ID","error":{"code":"ConstraintViolationError","message":"The request is malformed.","details":[{"code":"UnprocessableEntityError","target":"commands[0].arguments[0]","message":"commands[0].arguments[0]: integer found, string expected","details":[]}]}}
[21/02/2022, 15:19:51] [Samsung TV] The request is malformed.
[21/02/2022, 15:19:51] [Samsung TV] [DEBUG] undefined
2022-02-21T21:19:51.867Z HAP-NodeJS:Characteristic [On] Received error from set handler undefined
2022-02-21T21:19:51.868Z HAP-NodeJS:Accessory [Samsung TV C8DF] Error setting Characteristic "On" to value 0:  SERVICE_COMMUNICATION_FAILURE
2022-02-21T21:19:51.871Z HAP-NodeJS:EventedHTTPServer:Connection [::ffff:IP_ADDRESS] HTTP Response is finished
```

I went ahead and cleared all values in the "Channel" command and replaced them with old-style `"KEY_[NUMBER],KEY_ENTER"` commands.

Looking forward to your reply, thank you in advance!

**Comment by tavicu** [comment]
> * Didn't do much testing of the new "power-on delay" change, but as far as I could tell, using a HomeKit scene to power on the TV and switch inputs to HDMI2 is not working. The TV powered on, but it stayed in HDMI1 for me (I have an Apple TV box that powers on via CEC and forces the TV to HDMI1, not sure if this is affecting this feature in my case). I'll do more testing changing the new `wait_time` parameter

The CEC may be the fault here. Do you have any error in your console? Also when you run in debug mode, do you see the command to change the input to HDMI2, and right after that you should see the response from API.

Like this:
```
[2/22/2022, 9:44:45 AM] [Test TV] [DEBUG] https://api.smartthings.com/v1/devices/xxx/commands {
  commands: [
    {
      component: 'main',
      capability: 'mediaInputSource',
      command: 'setInputSource',
      arguments: [Array]
    }
  ]
}
[2/22/2022, 9:44:46 AM] [Test TV] [DEBUG] {"results":[{"id":"d4dfba71-7a1c-4389-abf5-eeafe286d17a","status":"ACCEPTED"}]}
```

> * New "set volume" feature is working, but only if the TV output is set to the TV speaker. My audio is configured to output using the HDMI ARC port; **can this feature be supported to set volume via HDMI?**

I just tested it and it's working on the HDMI also. The problem it appears to be on our side because the API accept only numbers and in your config you probably have `"volume": "22"`, please change it to `"volume": 22` and it should work until i release a new version which converts the value to number.

> * New "set channel" feature is not working, at least for my cable TV box via HDMI. Switches for my cable TV stations all have channel numbers and back in beta.4, the channels changed correctly. In beta.6, the presence of channel numbers in the "channel" parameter of the switch leads to instability, making the Home app report "not responding" status on my TV. Deleting the value from the "channel" parameter makes the plugin responsive again

I don't know if the SmartThings API supports changing channel on HDMI, it may work only on digitalTv! Unfortunately i don't have any cable box since everything on my side is directly by cable. In beta.4 it changed correctly because it was the hack which pressed the keys and that works probably because of Universal Control feature of TV (control HDMI device with your main TV remote).

Right now it's failing because the SmartThings API expects a string and you send a number. That means in your config file you have set `"channel": 22`. Please change it to `"channel": "22"` and let me know if it's working. The fix where i will convert all numbers to string it's on the way.

If it's not working with `"channel": "22"` then your option will be to use command like you said you did (`"KEY_[NUMBER],KEY_ENTER"`).

> * All switches created in this version "mis-behaved" in the Home app. For example, activating a switch to change channels to the BBC made switches for other channels come "on" at the same time (I believe this was due to every switch I configured having a value in the "input" parameter: HDMI2. Since the BBC is on HDMI2, all other switches with HDMI2 in the "input" parameter also got activated). In previous versions of the plugin, each switch in the Home app would act as a "toggle," that is, the switch would change to "on," the commands would be performed on the TV, and then the switch would go back to "off."

Indeed, having the `input` parameter on multiple switches will toggle all because in a switch there are only a few options that are stateful (`sleep`, `app`, `input`), all other options are stateless. 
So if you are having 5 switches that set input to HDMI and volume from 1 to 5. Once you are on HDMI input they will all be toggled because only the input will be checked and that will return `true` for all 5 switches.

What i could do here is add an option in config of the switch to force it to be stateless. That way even if the switch is stateful it will be forced to act as a stateless switch (switch back to off after toggle it)

**Comment by ramm64** [comment]
### Changing input to HDMI2 on TV power-on

> The CEC may be the fault here. Do you have any error in your console? Also when you run in debug mode, do you see the command to change the input to HDMI2, and right after that you should see the response from API.

Here are the console messages when I switched on the TV and tried to change to HDMI2:

```
[22/02/2022, 07:44:19] [Samsung TV] [DEBUG] https://api.smartthings.com/v1/devices/MY-TV-ID/commands {
  commands: [
    {
      component: 'main',
      capability: 'mediaInputSource',
      command: 'setInputSource',
      arguments: [Array]
    }
  ]
}
[22/02/2022, 07:44:19] [Samsung TV] [DEBUG] https://api.smartthings.com/v1/devices/MY-TV-ID/commands {
  commands: [ { component: 'main', capability: 'refresh', command: 'refresh' } ]
}
[22/02/2022, 07:44:19] [Samsung TV] [DEBUG] {"requestId":"9DAEA711-5FF8-4C7B-BA58-A33444DA9981","error":{"code":"ConflictError","message":"invalid device state","details":[]}}
[22/02/2022, 07:44:19] [Samsung TV] [DEBUG] {"requestId":"FC807ED9-30DD-4153-92BA-47348DE5BCE0","error":{"code":"ConflictError","message":"invalid device state","details":[]}}
```

### Volume issue
Changed volume value to a number (no quotations) and it still didn't work on HDMI for me. I've got a soundbar connected via ARC port and when I change volume with my remote, the TV screen shows an animation with "+" and "-" signs, and the volume changes. When I try to change volume with the plugin, nothing happens (no screen animation, no volume change).  Console output below:

```
[22/02/2022, 07:44:45] [Samsung TV] [DEBUG] https://api.smartthings.com/v1/devices/MY-TV-ID/commands {
  commands: [
    {
      component: 'main',
      capability: 'audioVolume',
      command: 'setVolume',
      arguments: [Array]
    }
  ]
}
[22/02/2022, 07:44:46] [Samsung TV] [DEBUG] {"main":{"mndt":{"value":"2021-01-01"},"mnfv":{"value":"T-KSU2EAKUC-1903.0"},"supportsPowerOnByOcf":{"value":"true"},"mnhw":{"value":""},"data":{"value":null},"di":{"value":"MY-TV-ID"},"supportedPlaybackCommands":{"value":"[\"play\",\"pause\",\"stop\",\"fastForward\",\"rewind\"]"},"supportedInputSources":{"value":"[\"digitalTv\",\"HDMI1\",\"HDMI2\",\"HDMI3\"]"},"mnsl":{"value":""},"supportedPictureModesMap":{"value":"[{\"id\":\"modeDynamic\",\"name\":\"Dynamic\"},{\"id\":\"modeFilmmakerMode\",\"name\":\"FILMMAKER MODE\"},{\"id\":\"modeMovie\",\"name\":\"Movie\"},{\"id\":\"modeNatural\",\"name\":\"Natural\"},{\"id\":\"modeStandard\",\"name\":\"Standard\"}]"},"dmv":{"value":"res.1.1.0,sh.1.1.0"},"error":{"value":null},"switch":{"value":"on"},"vid":{"value":"VD-STV-2021"},"mnpv":{"value":"6.0"},"supportedPictureModes":{"value":"[\"Dynamic\",\"FILMMAKER MODE\",\"Movie\",\"Natural\",\"Standard\"]"},"firmwareVersion":{"value":"3.4.0"},"icv":{"value":"core.1.1.0"},"supportedSoundModes":{"value":"[\"Standard\"]"},"st":{"value":"2021-12-10T19:27:02Z"},"pictureMode":{"value":"Movie"},"supportedSoundModesMap":{"value":"[{\"id\":\"modeExternalStandard\",\"name\":\"Standard\"}]"},"disabledCapabilities":{"value":"[\"samsungvd.ambient\",\"samsungvd.ambientContent\",\"samsungim.fixedFindNode\",\"samsungvd.supportsFeatures\"]"},"soundMode":{"value":"Standard"},"playbackStatus":{"value":null},"supportedTrackControlCommands":{"value":null},"mobileCamSupported":{"value":null},"mute":{"value":"unmuted"},"supportedAmbientApps":{"value":"[]"},"n":{"value":"Samsung TV"},"mnmo":{"value":"UN43AU7000FXZX"},"volume":{"value":"5","unit":"%"},"tvChannel":{"value":""},"mnmn":{"value":"Samsung Electronics"},"imeAdvSupported":{"value":null},"mnml":{"value":""},"supportedInputSourcesMap":{"value":"[{\"id\":\"dtv\",\"name\":\"TV\"},{\"id\":\"HDMI1\",\"name\":\"Apple TV\"},{\"id\":\"HDMI2\",\"name\":\"Megacable\"},{\"id\":\"HDMI3\",\"name\":\"HDMI 3\"}]"},"tvChannelName":{"value":""},"mnos":{"value":"Tizen"},"pi":{"value":"5157e3b6-cd92-cdb6-b527-9ff3be92f22b"},"inputSource":{"value":"HDMI1"}}}
[22/02/2022, 07:44:46] [Samsung TV] [DEBUG] {"results":[{"id":"XXXX","status":"ACCEPTED"}]}
2022-02-22T13:44:46.525Z HAP-NodeJS:EventedHTTPServer [::ffff:XXXX] Muting event '1.364' notification for this connection since it originated here.
2022-02-22T13:44:46.525Z HAP-NodeJS:Accessory [Samsung TV C8DF] Setting Characteristic "On" to value 1
2022-02-22T13:44:46.528Z HAP-NodeJS:EventedHTTPServer:Connection [::ffff:XXXX] HTTP Response is finished
```

### Change channel issue 

> Please change it to `"channel": "22"`

Tried adding quotations to the channel number and it didn't work, it did not change the channel. Console log below:

```
[22/02/2022, 07:45:33] [Samsung TV] [DEBUG] https://api.smartthings.com/v1/devices/MY-TV-ID/commands {
  commands: [
    {
      component: 'main',
      capability: 'tvChannel',
      command: 'setTvChannel',
      arguments: [Array]
    }
  ]
}
[22/02/2022, 07:45:33] [Hue HB] Philips hue: request 13118: GET /lights
[22/02/2022, 07:45:34] [Hue HB] Philips hue: request 13118: 200 OK
[22/02/2022, 07:45:34] [Samsung TV] [DEBUG] {"results":[{"id":"MY-TV-ID","status":"ACCEPTED"}]}
2022-02-22T13:45:34.178Z HAP-NodeJS:EventedHTTPServer [::ffff:192.168.0.19] Muting event '1.349' notification for this connection since it originated here.
2022-02-22T13:45:34.178Z HAP-NodeJS:Accessory [Samsung TV C8DF] Setting Characteristic "On" to value 1
2022-02-22T13:45:34.181Z HAP-NodeJS:EventedHTTPServer:Connection [::ffff:192.168.0.19] HTTP Response is finished
```

### Stateless vs. Stateful switches

> So if you are having 5 switches that set input to HDMI and volume from 1 to 5. Once you are on HDMI input they will all be toggled because only the input will be checked and that will return true for all 5 switches.

I see what you mean. However, I seem to recall that in previous betas (beta.4?), my switches remained stateless even with all of them having "HDMI2" in their input parameters. Maybe I am not remembering this correctly... (that could definitely be the case!)

Many thanks for all your hard work on this beta version! 🙏🏽

**Comment by tavicu** [comment]
> ### Changing input to HDMI2 on TV power-on
> 
> Here are the console messages when I switched on the TV and tried to change to HDMI2:

Set `wait_time` to `10000` (10 seconds). From the response the API see your TV as offline and don't accept requests.


> ### Volume issue
> Changed volume value to a number (no quotations) and it still didn't work on HDMI for me. I've got a soundbar connected via ARC port and when I change volume with my remote, the TV screen shows an animation with "+" and "-" signs, and the volume changes. When I try to change volume with the plugin, nothing happens (no screen animation, no volume change). Console output below:

Yup, now that you said about + and - it makes sense. Setting a specific volume will not work for speakers connected with HDMI. When the TV has speakers connected this way it knows only of Volume UP and DOWN.

So unfortunately this feature will not work for you.
Me for example, until i implemented this feature, i had an automation that set the volume to 2 (before going to sleep). I did that with commands like this `KEY_VOLDOWN*2s, KEY_VOLUP*2`. (2 seconds because it was enough, my volume is never more than 25)

_Nice to know it, i will add it in the "known problems"_

> ### Change channel issue
> 
> Tried adding quotations to the channel number and it didn't work, it did not change the channel. Console log below:

From the logs it appears that the SmartThings API accepted the channel command with success. Unfortunately this means that the API can't control the channel if you are on HDMI. 

So you will have to use the "hack" with commands.
Adding an option to use the hack with `commands` attribute it's not available since i want to migrate more features to SmartThings and in the end to remove the depreciated ones.

> ### Stateless vs. Stateful switches
> I see what you mean. However, I seem to recall that in previous betas (beta.4?), my switches remained stateless even with all of them having "HDMI2" in their input parameters. Maybe I am not remembering this correctly... (that could definitely be the case!)

That's because the feature to select a Switch based on the input was implemented in beta.4 (if i'm not wrong).

I worked today too and made some changes with switches that will improve the way it shows selected but the "problem" of having a stateful + stateless attributes will remain.

But i have implemented an option to force the switch to be stateless :)


> Many thanks for all your hard work on this beta version! 🙏🏽

Glad to see people enjoying it! To be fair i only use Turn ON/OFF, setVolume and sleep features on my TVs! So any other feature this plugin have was implemented for you guys :)

**Comment by tavicu** [comment]
New beta released (**v5.2.0-beta.7**)
For installing and updating follow the steps from here: https://github.com/tavicu/homebridge-samsung-tizen/issues/321#issuecomment-935886172

I tried to add a `debounce` for SmartThings API since it accepts a list of commands but it looks like SmartThings API don't work that good and not execute all commands. De debounce was a mechanism that collected all commands and send only one request to SmartThing instead of sending a request for each command. But as i said, the API don't execute some of the commands sometimes ... so i revert it (after i worked for hours ... :D)

What changed:

1. Changed the order of executing a switch. Now it will first change the input before executing the `channel` or other commands.
2. Custom switches will be stateful only if at least one of the attributes can be turned off. Right now only `sleep` can be turned off.
3. Removed the feature that list installed apps because Samsung removed the API with newer TVs and it also pushed an update for most of the older TVs. (not related to SmartThings API)
4. Add debug message that prints what info are saved in storage when the device intitialize
5. Add debug message that shows when the TV paired with success (and the token)

At the moment i think we are 99% finished an ready for release.

I know i said in a previous message that i will not release this version until the new documentation will be ready. But since i don't know how my time will be and when i will finish the new documentation i decided to just make a fast page in the wiki on how to set SmartThings and release it.

If no other problems appears with this beta this week, i think the release will be next week.

**Comment by mrMiimo** [comment]
if I understand correctly now it is also possible to change channel via API by setting a switch ... right?
Would it be possible to set it as INPUT? so I avoid having an endless list of switches
Thank you for these great, great developments!

**Comment by tavicu** [comment]
> if I understand correctly now it is also possible to change `channel` via API by setting a switch ... right?

You are right.

> Would it be possible to set it as INPUT? so I avoid having an endless list of switches

Nope. Inputs should be inputs and only that :). I'm sorry

**Comment by ramm64** [comment]
> New beta released (**v5.2.0-beta.7**) For installing and updating follow the steps from here: [#321 (comment)](https://github.com/tavicu/homebridge-samsung-tizen/issues/321#issuecomment-935886172)

Alright, new beta.7 is now installed. I also went back setting my channels with the `KEY_*` “hacks” and will continue using `KEY_VOL*` keys for volume adjustments.

Oh, and setting `wait_time` to `10000ms` successfully changed input to HDMI2 on TV power on.

I think I’m all set with this plugin! Thanks again for all your consideration and awesome work! 👏🏽

**Comment by mrMiimo** [comment]
> > Would it be possible to set it as INPUT? so I avoid having an endless list of switches
> 
> Nope. Inputs should be inputs and only that :). I'm sorry

you are right...
is that Homekit is sometimes so inconvenient, when you want to change what you see you always go to that "part of the screen" (at least me), having the list of favorite channels under the inputs makes it more convenient (at least for me).

My parents had an old LG TV natively homekit compatible which allowed them to do this.

Thanks again, it's a fantastic job (even in the betas) ✋🏼

**Comment by mrMiimo** [comment]
Is there any way to switch between apps (like netflix) or hdmi suorce directly to a tv channel?
This is my "input" configuration:
```
"inputs": [     
                {
		"type": "input",
		"name": "Apple TV",
		"value": "HDMI1",
		"wait_time": 2000
		},
		{
		"type": "app",
		"name": "YouTube",
		"value": "111299001912",
		"wait_time": 2000
		},
		{
		 "type": "app",
		 "name": "Prime Video",
		 "value": "3201512006785",
		 "wait_time": 2000
		},
		{
		 "type": "command",
		 "name": "Channel 1",
		 "value": "digitalTv,KEY_5,KEY_0,KEY_1,KEY_ENTER"
		},
		{
		 "type": "command",
		 "name": "Channel 2",
		 "value": "digitalTv,KEY_5,KEY_0,KEY_2,KEY_ENTER"
		}
          ]
```
When I try to change from the app directly to the channel, it's like the first `value` of the `command` line ("`digitalTv`") doesn't work.

**Comment by tavicu** [comment]
> Is there any way to change from apps (like netflix) or hdmi suorce directly to a tv channel? This is my "input" configuration:

Nope. Inputs are inputs, channels are channels :)
The only way is with a custom switch that change input and set channel

> When I try to change from the app directly to the channel, it's like the first `value` of the `command` line ("`digitalTv`") doesn't work.

Becaus this thing it's not implemented, don't know where you got it from.

**Comment by mrMiimo** [comment]
> Becaus this thing it's not implemented, don't know where you got it from.

oh, my fault ... sorry ✋🏼
Perhaps from someone's comment in some open "problem", forget it. Thanks 👍🏼

**Comment by schweppy** [comment]
Hi all,

I have searched for this issue but cant find anything. Just tried installing the beta for the first time to use inputs but getting the following fault. Any ideas? 

![C59EFF15-6BF0-4A8B-9D48-7A19F528A4FB](https://user-images.githubusercontent.com/17147950/156884877-2dc59392-921b-44fa-8780-7e31114f9e39.jpeg)

**Comment by MattRoe84** [comment]
Hi All

I also get an error:
[11/03/2022, 08:58:43] [Livingroom TV] The request is malformed.

Config:
 "inputs": [
                        {
                            "type": "app"
                        },
                        {
                            "type": "input",
                            "name": "Xbox",
                            "value": "HDMI1",
                            "wait_time": 2000
                        }
                    ],
                    "switches": [
                        {
                            "name": "Turn on Plex",
                            "power": true,
                            "app": "3201512006963"
                        },
                        {
                            "name": "Turn on Netflix",
                            "power": true,
                            "app": "3201907018807"
                        },
                        {
                            "name": "Turn on Xbox",
                            "input": "Xbox"
                        }
                    ]

**Comment by tavicu** [comment]
> I have searched for this issue but cant find anything. Just tried installing the beta for the first time to use inputs but getting the following fault. Any ideas?

That is the response from SmartThings API when it detects the TV as off. So your TV was off or your deviceId is wrong.


> I also get an error: [11/03/2022, 08:58:43] [Livingroom TV] The request is malformed.
> 
> Config: "inputs": [ { "type": "app" }, { "type": "input", "name": "Xbox", "value": "HDMI1", "wait_time": 2000 } ], "switches": [ { "name": "Turn on Plex", "power": true, "app": "3201512006963" }, { "name": "Turn on Netflix", "power": true, "app": "3201907018807" }, { "name": "Turn on Xbox", "input": "Xbox" } ]

1. `input: Xbox` it's not a valid input.
Possible values for now: digitalTv, USB, HDMI1, HDMI2, HDMI3, HDMI4, HDMI5, HDMI6

2. wait_time setting inputs are not available anymore: https://github.com/tavicu/homebridge-samsung-tizen/issues/321#issuecomment-1047050543

3. wait_time lower than `3000` will probably not work, don't know why changing it with a lower value than default

4. As i already told on every issue opened, without debug turned on i can't point specific problem because my crystal ball it's not working at the moment :)

**Comment by MattRoe84** [comment]
thanks @tavicu apologies as you can tell im quite new to this

**Comment by tavicu** [comment]
> thanks @tavicu apologies as you can tell im quite new to this

Sorry if i sounded rude but i'm trying to write it everywhere and it seems like I do something wrong because nobody see it :D 

I'm working on a new documentation which i hope it will solve this problem, or at least improve it. Because at the moment 99% of issues are just an error without context or debug logs which is very hard to catch it and try to solve it ...

**Comment by schweppy** [comment]
> > I have searched for this issue but cant find anything. Just tried installing the beta for the first time to use inputs but getting the following fault. Any ideas?
> 
> That is the response from SmartThings API when it detects the TV as off. So your TV was off or your deviceId is wrong.
> 
> > I also get an error: [11/03/2022, 08:58:43] [Livingroom TV] The request is malformed.
> > Config: "inputs": [ { "type": "app" }, { "type": "input", "name": "Xbox", "value": "HDMI1", "wait_time": 2000 } ], "switches": [ { "name": "Turn on Plex", "power": true, "app": "3201512006963" }, { "name": "Turn on Netflix", "power": true, "app": "3201907018807" }, { "name": "Turn on Xbox", "input": "Xbox" } ]
> 
> 1. `input: Xbox` it's not a valid input.
>    Possible values for now: digitalTv, USB, HDMI1, HDMI2, HDMI3, HDMI4, HDMI5, HDMI6
> 2. wait_time setting inputs are not available anymore: [[Feature] Samsung SmartThings API support for Input Source, etc. #321 (comment)](https://github.com/tavicu/homebridge-samsung-tizen/issues/321#issuecomment-1047050543)
> 3. wait_time lower than `3000` will probably not work, don't know why changing it with a lower value than default
> 4. As i already told on every issue opened, without debug turned on and steps you make to get the error i can't point specific problem because my crystal ball it's not working at the moment :)

Thank you. 

I just checked the ST app and the TV told me that it had a “new owner” so I reset with my ST account and now the plug-in is working correctly. Appreciate it!

**Comment by MattRoe84** [comment]
> > thanks @tavicu apologies as you can tell im quite new to this
> 
> Sorry if i sounded rude but i'm trying to write it everywhere and it seems like I do something wrong because nobody see it :D
> 
> I'm working on a new documentation which i hope it will solve this problem, or at least improve it. Because at the moment 99% of issues are just an error without context or debug logs which is very hard to catch it and try to solve it ...

Not rude at all, we appreciate your help, ive now amended the config and its now working perfectly, thanks

**Comment by tavicu** [comment]
I have released the feature for everybody :)

I also created a new page in the Wiki on how to setup SmartThings API and updated the Inputs and Switches pages with the new information.

I hope i will be ready asap with the new documentation so we can remove the wiki and have a more nicer instructions :)

https://github.com/tavicu/homebridge-samsung-tizen/wiki/SmartThings-Setup
https://github.com/tavicu/homebridge-samsung-tizen/wiki/Inputs
https://github.com/tavicu/homebridge-samsung-tizen/wiki/Switches

**Comment by tavicu** [comment]
I had two donations after releasing this version telling me about the awesome work for the SmartThings API integration.
If you are over here I just want to thank you!

It's not about the value, it's about feeling appreciated for the work I do and the time I dedicate for the community!
